### ISSUE TITLE
Use attributeCount=0 as sentinel instead of VertexBufferNotUsed?

### ISSUE THREAD

**Issue by kainino0x** [webgpu-native/webgpu-headers#432]
See discussion here: https://github.com/webgpu-native/webgpu-headers/pull/424#issuecomment-2495245707

> * to answer question: **No, don't need `Undefined`**
> * **KN to investigate using attributeCount (what does the JS API do - and could it change to treat `[]` as a hole?)**

But if we do remove VertexBufferNotUsed=0 then there will still be a reserved 0 for Undefined and then we should use it for defaulting since that's what we do everywhere else

**Comment by kainino0x** [comment]
Here is the relevant spec discussion: #2864

**Comment by kainino0x** [comment]
tl;dr: This was actually how webgpu.h in Dawn worked 2.5 years ago, but then we found that there was a mismatch between C and JS so we added `VertexBufferNotUsed`. We (the WebGPU editors/WG) just didn't think it was necessary to change the JS spec. Honestly though, the only reason not to change it in JS is that the otherwise-meaningless validation (of arrayStride) might catch some weird programming error (like you forgot to initialize your attributes array so it's empty, somehow).

If we do use `attributeCount=0` as the sentinel, but JS doesn't change, then:
- C-on-JS implementations will simply translate `WGPUVertexBufferLayout`s with `attributeCount=0` as `null`.
- JS-on-C implementations will have to inject an error (to `createRenderPipeline`) when they get `attributes=[]`. (This problem goes away if JS changes to treat it the same as a hole.)

**Comment by kainino0x** [comment]
> Honestly though, the only reason not to change it in JS is that the otherwise-meaningless validation (still validates `arrayStride`, and draw calls still validate there's a vertex buffer bound even though it's not really used) might catch some weird programming error.

What I wrote here was a little wrong before; edited.

**Comment by kainino0x** [comment]
Nov 25 meeting:

*   KN: Inclined to change C back, and then try to change JS to match (loosen the validation there so it doesn't check you have a vertex buffer bound if you're actually not using it)
*   CF: Agree, this seems silly
*   …
*   **KN to attempt a webgpu.h PR**

**Comment by kainino0x** [comment]
> * JS-on-C implementations will have to inject an error (to `createRenderPipeline`) when they get `attributes=[]`. (This problem goes away if JS changes to treat it the same as a hole.)

Sorry, this was totally incorrect, don't know why I said it. JS does not treat this as an error. Instead it requires there be a vertex buffer bound at that slot, even though it will not be used. I do not think there is a reasonable way we can use `attributeCount=0` as the sentinel unless either the JS API changes (https://github.com/gpuweb/gpuweb/issues/4999) or browsers communicate the distinction through some side-channel (extension, or some possibly-temporary sentinel value like whether the step mode is `Undefined` or not).

**Comment by kainino0x** [comment]
OK, here's a proposal for how we can do this in C without `VertexBufferNotUsed`: #439

**Comment by kainino0x** [comment]
Dec 9 meeting:

* KN: Turned out JS needs to distinguish `null` vs `{ attributes: [] }`. I suggested changing this in the WG, but they preferred not to introduce extra steps to the spec. Regardless of whether JS ever changes, came up with a proposal for how to do this in C: the sentinel value is `stepMode=Undefined` AND `attributeCount=0`.
  * Remove WGPUVertexStepMode\_VertexBufferNotUsed [\#439](https://github.com/webgpu-native/webgpu-headers/pull/439)  
  * CF: **Sounds good**  
  * (LK/KN discussion of migration in Dawn)
